typing/rec_check.ml

```diff
@@ -306,333 +252,841 @@ let classify_expression : Typedtree.expression -> sd =
+        let field_mode = match rep with
+          | Record_float -> Dereference
+          | Record_unboxed _ -> Return
           | Record_regular | Record_inlined _
-          | Record_extension -> Use.guard
+          | Record_extension -> Guard
         in
-        let field env = function
-            _, Kept _ -> Use.empty
-          | _, Overridden (_, e) -> expression env e
+        let field (_label, field_def) = match field_def with
+            Kept _ -> empty
+          | Overridden (_, e) -> expression e
         in
-        Use.join
-          (use (array field env es))
-          (option expression env eo)
+        join [
+          array field es << field_mode;
+          option expression eo << Dereference
+        ]
     | Texp_ifthenelse (cond, ifso, ifnot) ->
-        Use.(join (inspect (expression env cond))
-                (join
-                    (expression env ifso)
-                    (option expression env ifnot)))
+      (*
+        Gc |- c: m[Dereference]
+        G1 |- e1: m
+        G2 |- e2: m
+        ---
+        Gc + G1 + G2 |- if c then e1 else e2: m
+
+      Note: `if c then e1 else e2` is treated in the same way as
+      `match c with true -> e1 | false -> e2`
+      *)
+      join [
+        expression cond << Dereference;
+        expression ifso;
+        option expression ifnot;
+      ]
     | Texp_setfield (e1, _, _, e2) ->
-        Use.(join (inspect (expression env e1))
-              (inspect (expression env e2)))
+      (*
+        G1 |- e1: m[Dereference]
+        G2 |- e2: m[Dereference]
+        ---
+        G1 + G2 |- e1.x <- e2: m
+
+        Note: e2 is dereferenced in the case of a field assignment to
+        a record of unboxed floats in that case, e2 evaluates to
+        a boxed float and it is unboxed on assignment.
+      *)
+      join [
+        expression e1 << Dereference;
+        expression e2 << Dereference;
+      ]
     | Texp_sequence (e1, e2) ->
-      Use.(join (discard (expression env e1))
-              (expression env e2))
-    | Texp_while (e1, e2) ->
-      Use.(join (inspect (expression env e1))
-              (discard (expression env e2)))
+      (*
+        G1 |- e1: m[Guard]
+        G2 |- e2: m
+        --------------------
+        G1 + G2 |- e1; e2: m
+
+        Note: `e1; e2` is treated in the same way as `let _ = e1 in e2`
+      *)
+      join [
+        expression e1 << Guard;
+        expression e2;
+      ]
+    | Texp_while (cond, body) ->
+      (*
+        G1 |- cond: m[Dereference]
+        G2 |- body: m[Guard]
+        ---------------------------------
+        G1 + G2 |- while cond do body done: m
+      *)
+      join [
+        expression cond << Dereference;
+        expression body << Guard;
+      ]
     | Texp_send (e1, _, eo) ->
-      Use.(join (inspect (expression env e1))
-              (inspect (option expression env eo)))
+      (*
+        G |- e: m[Dereference]
+        ---------------------- (plus weird 'eo' option)
```

Comment:
It doesn't, because we aren't. The code must be safe because `Dereference` is the most conservative possible choice for `eo`, but I still have mostly no idea what this `eo` thing does (despite having looked at the compilation scheme). It's right that people hitting this case would be less-confident and would want (if they are changing it) to check things by themselves.